Docs update (#550)
sylt-lang/sylt-lang: docs/guide.adoc

Constant variables (apart from functions) should be written in
`SCREAMING_SNAKE_CASE`.

> The type of the variable is inferred but can be supplied.

Reviewer: ```suggestion
The type of the variable is inferred but can be supplied if you want to be explicit.
```

it in the constant variable `greet`. We also use the print-function in the
function body and print a string to the terminal.

> The function block starts with `do` and ends with `end`. The `do` can be
> replaced with a `->` if the function returns something. See the following
> examples.

Reviewer: ```suggestion
The function block starts with `do` or `->`, and ends with `end`. `do` indicates no return value.
An `->` means the function returns something. See the following
examples.
```
It read a bit wierd - like arrows were special cases.

Reviewer 2: ```
The function block starts with `do` or `->` and ends with `end`. `do` indicates
no return value while `->` means the function returns something. See the
following examples.
```

it in the constant variable `greet`. We also use the print-function in the
function body and print a string to the terminal.

The function block starts with `do` or `->` and ends with `end`. `do` indicates
no return value while `->` means the function returns something. See the

> following examples.

Reviewer 2: `do` doesn't require no return value. Although I can't seem to think of a good rewording.

Reviewer: I think the whole sentence has to be restructured

.Map and implicit `ret`
[source, sylt]
----

> squares :: [1, 2, 3] -> map' fn a -> a * a end

Reviewer: Yum

`new_player` creates a `Player` with a built in update-function that
only mutates the created `Player` blob. The `Player` blob is baked

> into the update-function as the constant `self`.

Reviewer: ```suggestion
into the update-function as the psuedo-variable `self`.
```
I feel like claiming `self` to be constant is a bit confusing, since you can change the fields (which you can do with constants, but I'd say it's not the main point)

Reviewer 2: pseudo-variable is on the other hand a bit vague. I'd remove it all together.

```
The `Player` blob is baked
into the update-function as `self`.
```

The example makes it clear how it can be used.

----
l := [1, 2, 3]

> l -> for_each' print

Reviewer 2: pog

l -> for_each' fn a: int do

>     // ...

Reviewer: ```suggestion
    // many statements
```
Felt this was clearer - idk

sum := 6

> points_str := points -> map' fn p: int -> str do

Reviewer: You do know this is still allowed?

Author: Yes, but I don't see the point of being inconsistent. I'd say we remove that syntax, but it's not hurting anyone.

Reviewer: Removing it is gonna be tricky - but I like the optionality of it. It might also be helpfull if we want to allow functions definitions over multiple lines. I can also see it being used as a stylistic approach when inlining with a return type:
```
add_one :: fn a -> int do a + 1 end
```
This reads clearer than:
```
add_one :: fn a -> int a + 1 end
```
If you want to explicitly write the return type.

    eat_bug: fn -> void,
}

new_spider :: fn x, y ->
    Spider {
        hp: 5,
        position: position,
        eat_bug: fn do
            self.hp += 1
        end,
    }
end

spider := new_spider(0.0, 0.0)
----

> // Leaving this here until we decide what happens with ducktyping
> //Notice that the `Spider` blob has many fields in common with the `Creature`
> //blob from before. This fact can be used to do something called
> //https://en.wikipedia.org/wiki/Duck_typing[duck typing], which effectively
> //allows us to assign `Spider` blobs to `Creature` variables.
> //
> //.Duck typing
> //[source, sylt]
> //----
> //creatures: [Creature] = []
> //creatures -> push' new_spider' 0.0, 0.0
> //creatures[0].hp -= 1   // Ok
> //creatures[0].eat_bug() // Type error
> //----
> //
> //This type error may seem counter-intuitive. We know that `creatures[0]` is a
> //`Spider` which has the `eat_bug` function, yet we are not allowed to call it.
> //What is really going on here? Once the `Spider` blob is put into the list
> //of type `[Creature]` some type information is lost, meaning, there is no
> //reliable way of knowing the true type of something in the list. `creatures` is
> //a list of `Creature`s and other blobs containing the same fields as `Creature`,
> //though every element is treated as if being of type `Creature`.
> //
> //To call the `eat_bug` function anyway, forced types may be used.
> //
> //.Force general blob into specific blob
> //[source, sylt]
> //----
> //spider: !Spider = creatures[0]
> //spider.eat_bug() // Ok, as long as creatures[0] is a Spider
> //----

Reviewer: ```suggestion
```
Ducktyping is dead as far as I'm concerned. There's no way to add it into the type-system as it is.
